kernel/src/engine/arrow_utils.rs

```diff
@@ -1,89 +1,1338 @@
+        ReorderIndex {
+            index,
+            transform: ReorderIndexTransform::Child(children),
+        }
+    }
+
+    fn new_none(index: usize) -> Self {
+        ReorderIndex {
+            index,
+            transform: ReorderIndexTransform::None,
+        }
+    }
+
+    fn new_missing(index: usize, field: ArrowFieldRef) -> Self {
+        ReorderIndex {
+            index,
+            transform: ReorderIndexTransform::Missing(field),
+        }
+    }
+
+    /// Check if this reordering requires a transformation anywhere. See comment below on
+    /// [`is_ordered`] to understand why this is needed.
+    fn needs_transform(&self) -> bool {
+        match self.transform {
+            // if we're casting or inserting null, we need to transform
+            ReorderIndexTransform::Cast(_) | ReorderIndexTransform::Missing(_) => true,
+            // if our children are not ordered somehow, we need a transform
+            ReorderIndexTransform::Child(ref children) => !is_ordered(children),
+            // no transform needed
+            ReorderIndexTransform::None => false,
+        }
+    }
+}
+
+// count the number of physical columns, including nested ones in an `ArrowField`
+fn count_cols(field: &ArrowField) -> usize {
+    _count_cols(field.data_type())
+}
+
+fn _count_cols(dt: &ArrowDataType) -> usize {
+    match dt {
+        ArrowDataType::Struct(fields) => fields.iter().map(|f| count_cols(f)).sum(),
+        ArrowDataType::Union(fields, _) => fields.iter().map(|(_, f)| count_cols(f)).sum(),
+        ArrowDataType::List(field)
+        | ArrowDataType::LargeList(field)
+        | ArrowDataType::FixedSizeList(field, _)
+        | ArrowDataType::Map(field, _) => count_cols(field),
+        ArrowDataType::Dictionary(_, value_field) => _count_cols(value_field.as_ref()),
+        _ => 1, // other types are "real" fields, so count
+    }
+}
+
+/// helper function, does the same as `get_requested_indices` but at an offset. used to recurse into
+/// structs, lists, and maps. `parquet_offset` is how many parquet fields exist before processing
+/// this potentially nested schema. returns the number of parquet fields in `fields` (regardless of
+/// if they are selected or not) and reordering information for the requested fields.
+fn get_indices(
+    start_parquet_offset: usize,
+    requested_schema: &Schema,
+    fields: &Fields,
+    mask_indices: &mut Vec<usize>,
+) -> DeltaResult<(usize, Vec<ReorderIndex>)> {
+    let mut found_fields = HashSet::with_capacity(requested_schema.fields.len());
+    let mut reorder_indices = Vec::with_capacity(requested_schema.fields.len());
+    let mut parquet_offset = start_parquet_offset;
+    for (parquet_index, field) in fields.iter().enumerate() {
+        debug!(
+            "Getting indices for field {} with offset {parquet_offset}, with index {parquet_index}",
+            field.name()
+        );
+        match field.data_type() {
```

Comment:
aside: This is a really large match... 200+ LoC if I count correctly. I wish there were better ways to split up such code in rust, but that's a bit trickier given that enum variants are not types and so can't be part of a helper method signature... I suppose helper methods could take the deconstructed enum variant content instead, but then you lose some semantic info.